### ISSUE TITLE
1.5.5: After conversion from 1.4, mixer action is active when it should be inactive 

### ISSUE THREAD

**Issue by RC-SOAR** [FrSkyRC/ETHOS-Feedback-Community#3801]
**X20S, Ethos 1.5.5 nightly #81c6ae1f**

I believe this should be classed as a critical ìssue. It was reported by a user of my sailplane templates - he had converted the (working) setup from 1.4.17 to 1.5.4. 

The symptom was both ailerons moving up together when a right aileron command was applied.

Here's minimalist setup to reproduce the issue in 1.5.5 nightly:

1. Create a free mix, 
2. Set input = Aileron
3. Set 1st action Condition='Top', weight = -50%
4. Set 2nd action Condition='Bottom', weight = -100%
5. Apply right aileron 

_Expected:_ 'Top' action only should be active
_Actual_: 'Top' and 'Bottom' actions are both active

The issue also occurs with an elevator input. 
Conditions of 'Left' and 'Right' work as expected.

The incorrect response is reflected in the graph:

![IMG_3299](https://github.com/FrSkyRC/ETHOS-Feedback-Community/assets/6317479/247dffc2-826b-4fbc-99b4-9af033f2d94b)

**Comment by RC-SOAR** [comment]
I was puzzled as to why this issue was not identified earlier, and I think I have found the reason. It seems that the way mixer weights in 1.4 are converted to 1.5.x depends on the version.

In early versions of 1.5.x, up/down weights were translated to 'Right' and '!Right' - these work as expected. 
In later versions, up/down is translated to 'Top' and 'Bottom', and this seems to be the cause of the issue.

[EDIT] this appears to be related to #3639.

[EDIT2] Some additional info about the original context:
- The mixer source was a channel (not a stick). 
- The up/down weights in the original setup are both negative.
> FreeMix source=CH34:'AileronWithRate'
> Weight-up=−CH17:'AilDown'
> Weight-down=−100%

**Comment by bsongis-frsky** [comment]
Why don't you use Left / Right instead of Top / Bottom?

**Comment by bsongis** [comment]
Indeed if you apply a negative weight to a value above 0 in the "Top" line, it will go below 0 and then the "Bottom" will apply too. 
You should use Left / Right here!

**Comment by RC-SOAR** [comment]
Yes, I understand now (after reading a little more about top/right/bottom/left). However I had no choice as 'top' and 'bottom' were inserted during the migration from 1.4.17 to later versions of 1.5.x.

Please ignore my original example. Below are the actual migration paths of the affected mixer:

Original mixer in 1.4.17:
```
FreeMix source = CH34:'AilWithRate'
Weight-up = −CH17:'AilDown'
Weight-down = −100%
```

After flashing to 1.5.0 translates to this, which works correctly
```
FreeMix source = CH34:'AilWithRate'
Positive = −CH17:'AilDown'
!Positive = −100%
```

**However, flashing from 1.4.17 to 1.5.5 in one step translates to this, which results in unexpected behaviour, note use of top/bottom**
```
FreeMix source = CH34:'AilWithRate'
Top = −CH17:'AilDown'
Bottom = −100%
```

Summary of results:
1.4.17 -> 1.5.0 good.
1.4.17 -> 1.5.5 bad

In other words, whether the mixer is translated correctly depends on the minor version of the 1.5 target. This appears to be a migration issue.

**Comment by bsongis-frsky** [comment]
Indeed it is a conversion bug

**Comment by bsongis-frsky** [comment]
Indeed it is related to #3639 
The only solution I can see is to have this weight defined in only one action, like this

![image](https://github.com/FrSkyRC/ETHOS-Feedback-Community/assets/50834778/9d58cb50-a132-4d5b-94ce-c4c6d77c221b)

**Comment by bsongis-frsky** [comment]
After applying another fix so that the above combinaison works ...

![image](https://github.com/FrSkyRC/ETHOS-Feedback-Community/assets/50834778/43af349c-3754-4bd4-bae0-822610b5ba91)

**Comment by RC-SOAR** [comment]
I can see that the solution works for this particular case. 
However I'm not sure that two sequential actions can deal with the general case where weight-up and weight-down may both be supplied by channels with unkonwn value, and where the input may be subject to an arbitrary curve.

**Comment by bsongis-frsky** [comment]
2 sequencial actions won't work.
Here we need only one action, which deals with 2 different weights, one for positive input, the other for negative input
LEFT / RIGHT => depends on the very first input (Aileron in your example, throttle in scenario #3639)
TOP / BOTTOM => depends on the current input (Aileron in your example, Curve output in scenario #3639)

The conversion must use TOP / BOTTOM

![image](https://github.com/FrSkyRC/ETHOS-Feedback-Community/assets/50834778/e4192cb7-96c7-4ab2-a748-c922d595d71f)

**Comment by RC-SOAR** [comment]
Thanks, this looks like a neat solution.
